src/v/storage/segment.cc

```diff
@@ -202,15 +203,35 @@ ss::future<> segment::release_appender() {
                      a = std::move(a),
                      c = std::move(c),
                      i = std::move(i)]() mutable {
-                        return write_lock().then(
-                          [this,
-                           a = std::move(a),
-                           c = std::move(c),
-                           i = std::move(i)](ss::rwlock::holder h) mutable {
-                              return do_release_appender(
-                                       std::move(a), std::move(c), std::move(i))
-                                .finally([h = std::move(h)] {});
-                          });
+                        return ss::repeat([this,
+                                           a = std::move(a),
+                                           c = std::move(c),
+                                           i = std::move(i)]() mutable {
+                            /**
+                             * we have to use 'spin lock' approach in here singe
+                             * we do not want to make
+                             */
+                            if (_destructive_ops.try_write_lock()) {
+                                _destructive_ops.write_unlock();
+                                return write_lock()
+                                  .then([this,
+                                         a = std::move(a),
+                                         c = std::move(c),
+                                         i = std::move(i)](
+                                          ss::rwlock::holder h) mutable {
+                                      return do_release_appender(
+                                               std::move(a),
+                                               std::move(c),
+                                               std::move(i))
+                                        .finally([h = std::move(h)] {});
+                                  })
+                                  .then([] { return ss::stop_iteration::yes; });
+                            }
+                            // Notice that we do not need abortable sleep in
+                            // here since wait time is really short
+                            return ss::sleep(std::chrono::milliseconds(5))
```

Comment:
I think you can use `ss::later()` instead of setting a timer--but take a look at its implementation I don't exactly know what it does. what you seem to want here is a method to yield, and there are also other reactor interface that seem to allow a preempt to be requested.

Reply:
`ss::later()` is actually the way to break tail recursion. It indeed yield but it creates a task in a task queue, in this situation it would make unnecessary pressure with empty tasks in a queue.

Reply:
Doesn't sleeping add a task into the task queue (the future that resolves when the sleep is done?)